Add CONTRIBUTING.md (#99)
JOJ0/synadm: CONTRIBUTING.md

<!-- omit in toc -->
# Contributing

- [Types of Contribution](#types-of-contribution)
  - [Non-Programming](#non-programming)
  - [Programming](#programming)
- [Submitting Your Work](#submitting-your-work)
- [Expectations Regarding Timelines](#expectations-regarding-timelines)
- [Discuss \& Design First](#discuss--design-first)
- [Getting the Source \& Installing](#getting-the-source--installing)
  - [Install into a Virtual Python Environment](#install-into-a-virtual-python-environment)
  - [Install in Editable Mode](#install-in-editable-mode)
- [Command Design](#command-design)
  - [Implementation Examples](#implementation-examples)
  - [Helpers \& Utilities](#helpers--utilities)
  - [Logging](#logging)
  - [Code Documentation](#code-documentation)
  - [Code Style](#code-style)
- [Maintainers Notes](#maintainers-notes)
  - [Release process](#release-process)
- [Maintainership](#maintainership)

## Types of Contribution

No matter if you're a programmer or not there are many ways to contribute to the `synadm` project.

### Non-Programming

> * [Set up synadm](README.md#installation) and report whether the installation, configuration and update processes work as described or could be improved. We don't have the resources to test on many operating systems, thus a quick GitHub issue telling us: "It even works on Fancy-OS" would help already.

Reviewer: > We don't have the resources to test on many operating systems, thus a quick GitHub issue telling us: "It even works on Fancy-OS" would help already.

Not quite a good idea, it encourages to spam the issue tracker with non-issues. I think it's less worse if it's thrown into the room for synadm instead.

```suggestion
* [Set up synadm](README.md#installation) and report whether the installation, configuration and update processes work as described or could be improved. We don't have the resources to test on many operating systems, but you can [let us know](https://matrix.to/#/%23synadm%3Apeek-a-boo.at?via=jacksonchen666.com&via=maclemon.at&via=matrix.org&via=raim.ist) if it does work.
```

Author: Agree, I just had that in since the first versions where I was not sure if that damn thing works for anyone, so yeah we are past that long ago. Like your suggestion! Nice!

  * It's ok to open a Draft PR even if you don't want our feedback yet but it helps you in any way. We won't bother you until you ping us.
* Read the [Synapse Admin API docs](https://github.com/matrix-org/synapse/tree/master/docs/admin_api), pick a feature, implement it and send a pull-request. Don't forget to check if the feature is listed in the [Implementation Status / Commands List](README.md#implementation-status--commands-list) chapter already. if not, please include the addition to the list in your PR (separate commit if possible).

## Submitting Your Work

There are several ways to submit your work.

- Clone the repo, create a feature branch and submit a pull-request.
  - Consider enabling the [Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) option. We will deal with this permission responsibly. It proved to be a useful option for working together in the past and might even speed up getting a PR merged.

> - If you're not familiar with using git or just want to submit a smaller change (like a correction to the docs), use the “Edit file” button in the upper-right while viewing a file directly on the GitHub web interface. You can automatically send us a pull request for your changes.

Author: I would like to leave it in, since exactly this way encourages people to send quick fixes. I often got things like 3 spelling corrections. It's supertedious if people open issue for that and tell us "here and there are 3 wrong words". I'm never doing such tiny "medium important" things in the end... This quick edit feature is just superhelpful for that. Especially for people that don't know git or have never done a pull request. It's blocking them from acting at all I think.

> - Next to the "Edit file" button you'll find another option called "Open with github.dev" which opens a "Visual Studio Code" editor within your browser.

Author: I tried out this feature once and it was superslow, yeah happy to kick it actually. Specific to github is ok since the edit button is actually also specific to github, well probably gitlab and gitea do also have such a button. Anyway, we can remove it :-)

## Expectations Regarding Timelines

We are maintaining `synadm` in our spare time and currently are not sponsored by a company to do that. Nevertheless we try to be responsive to contributors and try to release new features as soon as they are merged.

> - We try to give a first review to your submitted pull-requests **within a week**, often you'll get an answer much faster though.

Author: Or I'll add that it's not a promis but rather a "rough estimate".

The message I'd like to bring across the net is: We try to not let things lye around for 3 weeks until we answer, if we can! A week is a goal but we can't always promise, especially in times of holiday and stuff.

We _try_ to give a first review......  is the wording....

> - We don't expect you to finish the PR / submit the changes we suggested "in time" but would like to ask you to **communicate your personal timeline**.

Author: Yeah that sounds more like what I was trying to say! Thanks!

- Usually it's not critical for a feature to be finished within a time frame. If you inform us that you will be able to finish your work within a month, but not right now, it's fine, **please just let us know**!
- If it's a serious bug we might ask you for permission to take over your branch and continue the work ourselves.

> - Unfortunately we often see contributors submitting PR's, correcting a few of our suggested changes but then disappear. Please don't do that! If you realize you won't find the time at all to continue your work, **please just let us know**. It's perfectly fine and there won't be hard feelings. If we consider your feature useful we might even take over and finish it ourselves.

Author: Well, yeah thanks for just shortening that one. That's the actual message. Leave out the clutter.

- We try to release merged features as soon as possible. Even the tiniest feature deserves a release. Sometimes we just can't make it soon enough because we don't find the time or we wait for something else because it makes sense to be released together. If a feature you submitted is merged and you'd need it urgently to be available on PyPi, **please kindly remind us** and we'll try to find the time.

## Discuss & Design First

It proved to be useful in the past if loose (feature) ideas would be discussed in #synadm:peek-a-boo.at first, which helps forming a more concrete idea that can than further be summarized into a well written github issue and finally be implemented by a `synadm` contributor or a `synadm` maintainer.

- If you'd like someone else to pick up your feature idea because you are not able to code it yourself, open a detailed GitHub issue describing the feature.

> - If you'd like to code it yourself, a separate issue is not required, just describe your feature within the PR.

Reviewer: Document, maybe that's the word

> - If you'd like to code it yourself, a separate issue is not required, just describe your feature within the PR.

Author: Just include a description of your feature within the PR.

## Getting the Source & Installing

### Install into a Virtual Python Environment

> To install `synadm` into its own private virtual Python environment, FIXME

Author: Whoops!

<!-- omit in toc -->
#### 1. Check Python Version

`python3 --version` should show at least v3.6.x

<!-- omit in toc -->
#### 2. Clone Repo:

```
git clone https://github.com/joj0/synadm
```

<!-- omit in toc -->
#### 3. Set up a Python3 Virtual Environment

Create and activate a virtual environment using the python3 venv module:

```

> python3 -m venv ~/.venvs/synadm

Reviewer: Personally I wouldn't do it that way. Instead, I would do:
```
python3 -m venv venv
```
This creates a virtual environment wherever your current directory is (hopefully within the synadm folder), so when you delete a project folder, the things related to it also gets deleted.

However, because the `.gitignore` file does not ignore `venv/` but only `env`, I have to do this:
```
python3 -m venv env
```
Still located within the project, just named differently.

Second point, the virtual environment is hidden away. So when the user is trying to figure out what to delete because their drive space is close to full, they might miss that hidden folder located at home.

Author: It's questionable if the several options of creating and managing virtualenvs and diskspace should be included in that doc at all but I always have the feeling that people that "halfway know" about how to install a pip package later wonder why they mess up things when they install everything in there user space, even things they tinker with... That's why I tend to keep a tutorial like that in my project repos sometimes.

I personally hate when venvs are directly to be found in project repos. I want to have all of them in one place. Sometimes I just delete bunches of them when I realize that was some experiment I'll never use again. The fact that a user has problems with hidden vs. non-hidden files on a Unix system I would not consider too much or if I'm being rude I'd say: Not my problem! :-))))

I could imagine to at least make it non-hidden. Yeah good idea. And also it would not make this tiny tutorial Unix-only. (Well slash vs. backslash would still be an issue if someone would code on windows...)

source ~/.venvs/synadm/bin/activate
```

Once your Python virtual environment is loaded and your prompt looks similar to `(synadm) ....  $ `, install directly into the environment:

<!-- omit in toc -->
#### 4. Install

```
cd synadm
pip install .
```

*Note: Don't forget to activate the venv when coming back to using `synadm` after a fresh login: `source ~/.venvs/synadm/bin/activate`*

<!-- omit in toc -->
#### 5. Run

> As long as your venv is loaded `synadm` should run fine without having to add a path in front of it

Reviewer: ```suggestion
As long as your venv is loaded, `synadm` should run fine without having to add a path in front of it
```

```
synadm -h
```

> ### Install in Editable Mode

Reviewer: This is more preferable for development, not the [former](https://github.com/JOJ0/synadm/pull/99/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R106).

Author: Sure,  I think that's what I tried to say here. I wanted to have the regular non-editable option in the turial as well since I figure this is the way most _users_ would install if they want to install the current master version.

Author: (and I think I link to this section from the main README IIRC), it used to be called "Install from git clone" or something... Maye now with the restructuring it doesn't make 100% sense any more...

@JacksonChen666: Question/Your opinion: I had 3 possible ways of installing synadm in the README:

- pip install synadm
- git clone..; pip install .
- git clone...; pip install -e .

Should we just remove the middle one?

If someone would like to install from git latest, he could as well install using `-e´

Would shorten a tiny bit...

Reviewer: > Should we just remove the middle one?

sure

> To be able to instantly be able to test the code you've been working on it makes sense to install from your git clone in _editable mode_

Reviewer: ```suggestion
To be able to instantly be able to test the code you've been working on it makes sense to install from your local clone of `synadm` in _editable mode_
```

Author: Yeah I tried to make that clear here that it's for editing/developing

and again it needs a backend method in `api.py`:
https://github.com/JOJ0/synadm/blob/107d34b38de71d6d21d78141e78a1b19d3dd5379/synadm/api.py#L72

### Helpers & Utilities

You'll find a couple of helpers & utilities [near the top of the api module's code](https://github.com/JOJ0/synadm/blob/master/synadm/api.py#L125), right below the `query()` method, within the `ApiRequest` class. For example we already provide methods to translate unix timestamps to human readable formats and vice versa.

If you need to defer code to a helper function because you require reusing it or otherwise think it's a cleaner approach, put it where you need it: Either as a subfunction in your backend method in the `synadm/api` module or in the frontend function in `synadm/cli/yourcommand`. If you think you've wrote a function that is very generic and might be useful to other `synadm/api` methods as well, put it next to the helpers in the `ApiRequest` class and tell us about it in a PR-comment.

### Logging

> Simply use `self.log.warn()` (or `.error/.debug/.info`) within the `synadm/api` module.

Reviewer: ```suggestion
Simply use `self.log.warn()` (or `.error`/`.debug`/`.info`) within the `synadm/api` module.
```

> From a command's frontend code in the `synadm/cmd` package, logging is available via the `helper` object that is available to every subcommand function, for example use `helper.log.warn()`.

Reviewer: There is no `./synadm/cmd`.

```suggestion
From a command's frontend code in the `synadm/cmd` package, logging is available via the `helper` object that is available to every subcommand function, for example: `helper.log.warn()`.
```

Author: Ah crap, I saw that error already but forgot. yeah should be `synadm/cli` I think.

We use the [google docstring format](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings).

_Not all parts of `synadm` are populated with such docstrings yet but we require it for newly written code._

### Code Style

We try to follow Python's [PEP 8](https://www.python.org/dev/peps/pep-0008/) guidelines as good as possible. We have an
automatic linter in place in our CI pipeline which runs whenever code is pushed to a branch. You'll notice a check going green or red at the very bottom when viewing your open PR. The check is called "Lint with flake8". If after a `git push` you notice the check gone red, click on it and you'll see details of what the linter detected you should correct.

`flake8` can be used directly on your code as well. Install it to your dev-environment:

```
pip install flake8
```

Run with:

`flake8 the_file_being_worked_on.py`

or leave out the filename to run it on all files and subdirectories.

Some more style and coding requirements:

> - Pythonism: `synadm` is a Python project and as such tries to use some of the very handy features of the language. Some  of those might not be familiar to you since often we realize that contributors are professional programmers used to other languages or even are admins who are rather at home on the shell than in Python code. That is certainly all fine and we are trying to support you with refactoring your code to become more pythonic.

Reviewer: ```suggestion
- Pythonism: `synadm` is a Python project and uses some handy features of the language. They might not be familiar to you since often we realize that contributors are professional programmers used to other languages or even are admins who are rather at home on the shell than in Python code. That is certainly all fine and we will support you with refactoring your code to be more pythonic.
```

> since often we realize that contributors are professional programmers used to other languages or even are admins who are rather at home on the shell than in Python code.

That is very many words and I don't know how to reduce it.

Author: Haha, then the words have won! LOL

- That said, we would still like `synadm` to stay approachable to Python beginners and would encourage contributors to try not to overcomplicate things.

> - We require a line length of 79 characters in Python code.

Reviewer: (Off-topicish) Why specifically 79? Because that means the code is at most 79 characters long, which is 1 off from 80.

Author: PEP8 requirese 79. Not true? Or am I misunderstanding your comment?

Reviewer: > PEP8 requirese 79. Not true?

I don't know. I don't feel like checking either.

> Or am I misunderstanding your comment?

Probably yes. My question is why 79 characters at most for 1 line instead of 80 characters?

Author: Sorry my bad, I understand now I also recall now and I chekcked: Yes, PEP 8 suggestion definitely is 79: https://peps.python.org/pep-0008/

I think I had this misunderstanding once myself and thought that 80 is the pep 8 suggestion. I realized it once I started to use editors and linters that have a "default pep 8" understanding, that it's 79. I then decided that it's best to follow this suggestion of 79 instead of having to reconfigure such tools as well as notify contributors that a non-pep8 default of 80 is used.

>   - Some editors have handy commands to quickly reformat multi-line text (e.g help texts of options). For example in `vim`, lines can be marked using `shift+V` and then reformatted by typing `gq`

Reviewer: ```suggestion
  - Some editors (e.g. [vim](https://vimhelp.org/change.txt.html#gq)) have handy commands to quickly reformat multi-line text (e.g help texts of options).
```

  - For code passages, staying within this boundary and at the same time keeping code well readable, this can be achieved using PEP8's many [line continuation suggestions](https://peps.python.org/pep-0008/#code-lay-out).
- We don't require a maximum line length in Markdown files, for example _this document_ or `README.md`. Please omit line-breaks in sentences, paragraphs, and so on, when editing these docs.

> - If you'd like to tidy up your commits or reduce commit count we welcome the use of `git commit --amend` and `git rebase`. If you'd like us to test your code or work together within your feature branch, be aware of that force pushing might complicate things and should rather be omitted in that case.

Reviewer: ```suggestion
- If you'd like to tidy up your commits or reduce commit count we welcome the use of `git commit --amend` and `git rebase`. If you'd like us to test your code or work together within your feature branch, be aware of that force pushing might complicate things and should be avoid in that case.
```

Author: avoidED you wanted to write?

Reviewer: > avoidED you wanted to write?

yes

## Maintainers Notes

> This section is a checklist for maintainers of the `synadm` project. Still there might be helpful information applying to contributors as well, thus we made it public.

Reviewer: "made it public" or "kept it public"?

```suggestion
This section is a checklist for maintainers of the `synadm` project. This information may be useful for contributors, so we kept it public.
```

Author: I don't know. kept it sounds good. just a feeling

> - We keep a branch named `dev` consistently existing that may be used for random things. If this branch is in use because a PR was opened from it, it's occupied and can't be used by others.

Reviewer: ```suggestion
- We keep a branch named `dev` consistently existing that may be used for random things. If a PR is opened from `dev`, it cannot be used by others.
```

> - Speaking of branches, we are not required to create branches in our own forks, we can just create them directly within the `synadm` repo. Checking out branches for e.g testing a feature in development is much easier than cloning and checking out from another maintainers fork.

Reviewer: ```suggestion
- Collaborators are not required to create branches in their own fork. Checking out branches already on the [JOJ0/synadm](https://github.com/JOJ0/synadm) repository is much easier than cloning and checking out from another maintainers fork.
```

It's definitely possible to checkout another repo:
```
git remote add <remote_name> <URL>
git fetch <remote_name>
git checkout <whatever>
```

It's actually what I do for my local clone (only the second and third step, first is only necessary per repo/remote).

Author: Sure, its just a little less things to do when a branch is directly in the same repo. I don't see a downside of having everythin in one repo and not having to switch remotes. What do you think?

(my wording was "much easier", not "impossible" :-))

Author: One more thing why I think it's more convenient. After some time I tend to get several remotes from several maintainers and need to clean up regularily. At least the branches that don't have a special remote I can rely on are our own owns and I don't need to fetch any particular remote, it will be done with a simple `git fetch`, but yeah again: Wording is "Collaborators are not required..." is the wording. We can still do it when there is good reasons. One thing where it's nice to have one's own fork are github acitons automations, it helps to know you would not mess with the automation of the core repo itself.

> - One person should be the "owner" of PR's opened by contributors and should request assistance by other maintainers if required. The ultimate decision to merge should be with the "PR owner". Who the owner is can be arranged in the _synadm maintainers chatroom_ but usually would be the person who initially reviews the submission.

Reviewer: ```suggestion
- One person should be the "owner" of PR's opened by contributors and should request assistance by other maintainers if required. The ultimate decision to merge should be with the "PR owner". Who the owner is can be arranged in the _synadm maintainers room_ (private) but usually would be the person who initially reviews the submission.
```

### Release process

Releasing a new version of `synadm` requires the following steps:

- Increase the version in all required files using the `bump.sh` script. `pip install bump2version` to your dev-environment to install the required tool. The final step to push to the master branch will be suggested by the script ('git push --follow-tags`).
- Following this push command the _release CI pipeline_ will be triggered and a tag and release draft be created on GitHub.
- Edit the generated draft and add release notes with the help of the "Generate release notes" button, but reformat into sections to be consistent with prior releases. If things were added without a PR, add those manually to the list of new features or fixes.
- Once the release notes are fine, make them public. The package can now be pushed to PyPi using the script `pypi.sh`
- Activate a new documentation version via [https://readthedocs.org/projects/synadm/versions/](https://readthedocs.org/projects/synadm/versions/), thus we provide a version of the online documentation consistent with the content of to the newly released `synadm` package.
- Spread the news in #synadm.peek-a-boo.at.

## Maintainership

> If you feel like supporting the `synadm` project on a regular basis, consider applying for maintainership. We believe in Open Source software and think no matter if you are a programmer or not, there is plenty of things to do and still a lot of room for improvement. Contact either via #synadm:peek-a-boo.at or private message to @jojo:peek-a-boo.at.

Reviewer: ```suggestion
If you feel like supporting the `synadm` project on a regular basis, consider applying for maintainership. We believe in Open Source software and think no matter if you are a programmer or not, there is plenty of things to do and still a lot of room for improvement. Contact either via [#synadm:peek-a-boo.at](https://matrix.to/#/%23synadm%3Apeek-a-boo.at?via=jacksonchen666.com&via=maclemon.at&via=matrix.org&via=raim.ist) or private message to [@jojo:peek-a-boo.at](https://matrix.to/#/@jojo:peek-a-boo.at).
```

> We believe in Open Source software

A bit too much PR-like tone for me

Author: PR as in public relations? LOL
